Implementation of `dropcollinear` feature in `GeneralizedLinearModel`

mousum-github opened the pull request:
This pull request supersedes #340 

We are looking for the `dropcollinear` feature in GLM and found #340. The PR, which opened in Oct'2019, looks very close to complete.

Apart from the changes covered in [PR 340](https://github.com/JuliaStats/GLM.jl/pull/340), we have added 

- A test case to compare LM with equivalent GLM
- Two test cases to match output with R
- Another implementation of `dof` for `CholeskyPivoted` since the `dof` will be calculated based on rank, instead of the number of coefficients. 
- update API documentation: Added `dropcollinear` in the list of keyword arguments. 

We have performed logistic regression with 10 independent variables, 10,000 rows and `dropcollinear` feature `true` and `false`; the dataset does not have a multicollinearity issue. The machine configuration is as follows: -
OS – Linux,
CPU - AMD Ryzen 5 3600 6-Core Processor
RAM – 64 GB

With `dropcollinear` = `true`: average time 4.381 ms 
With `dropcollinear` = `false`: average time 4.396 ms

nalimilan commented on src/glmfit.jl:
```diff
@@ -439,6 +443,7 @@ const FIT_GLM_DOC = """
     for a list of built-in links).
 
     # Keyword Arguments
+    - `dropcollinear::Bool=false`: Controls whether or not lm accepts a model matrix which is less-than-full rank.
```
Let's use the full description as for `?lm`:
```suggestion
    - `dropcollinear::Bool=false`: Controls whether or not `lm` accepts a model matrix which
      is less-than-full rank. If `true` (the default), only the first of each set of
      linearly-dependent columns is used. The coefficient for redundant linearly dependent columns is
      `0.0` and all associated statistics are set to `NaN`.
```

nalimilan commented on src/glmfit.jl:
```diff
@@ -475,6 +480,7 @@ function fit(::Type{M},
     y::AbstractVector{<:Real},
     d::UnivariateDistribution,
     l::Link = canonicallink(d);
+    dropcollinear::Bool = false,
```
For consistency with `lm` this should be `true` by default. Then we should also test that passing `dropcollinear=false` throws an error when columns are collinear.

mousum-github replied:
Okay. We also need to update existing examples in `glm`  where it prints the model.

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
     end
 end
+
+@testset "dropcollinearity in GLM" begin
```
```suggestion
@testset "dropcollinear with GLMs" begin
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+    dfrm[!, :x2] = randn(StableRNG(456), num_rows)
+    dfrm[!, :x3]= 2*dfrm[!, :x1] + 3*dfrm[!, :x2]
+    dfrm[!, :y] = Int.(randn(StableRNG(9999999), num_rows) .> 0)
+    @testset "General test" begin
```
This testset isn't very useful as it doesn't check many quantities (coefficients, etc.) like the ones below. Better drop the first model completely as it's tested below. The second one should either be dropped or tested against R (and put below).

Also, can you add line breaks before `@testset`?

mousum-github replied:
Okay - we will change and update here

mousum-github replied:
I have updated the test cases.

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+        
+        f1 = @eval(@formula(y ~ 1+x1+x2+x3))
+
+        m1 = fit(GeneralizedLinearModel,
+                f1,
+                dfrm,
+                Binomial(),
+                GLM.LogitLink();
+                dropcollinear=true)
```
```suggestion
        m1 = fit(GeneralizedLinearModel,
                 f1,
                 dfrm,
                 Binomial(),
                 LogitLink();
                 dropcollinear=true)
```
(Same below.)

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+@testset "dropcollinearity in GLM" begin
+    num_rows = 100_000
+    dfrm = DataFrame()
+    dfrm[!, :x1] = randn(StableRNG(123), num_rows)
+    dfrm[!, :x2] = randn(StableRNG(456), num_rows)
+    dfrm[!, :x3]= 2*dfrm[!, :x1] + 3*dfrm[!, :x2]
+    dfrm[!, :y] = Int.(randn(StableRNG(9999999), num_rows) .> 0)
```
```suggestion
    dfrm.x1 = randn(StableRNG(123), num_rows)
    dfrm.x2 = randn(StableRNG(456), num_rows)
    dfrm.x3= 2*dfrm.x1 + 3*dfrm.x2
    dfrm.y = Int.(randn(StableRNG(9999999), num_rows) .> 0)
```

nalimilan commented on src/glmfit.jl:
```diff
@@ -273,6 +273,10 @@ end
 
+dof(obj::GeneralizedLinearModel{<:GlmResp,<:DensePredChol{<:Real,<:CholeskyPivoted}}) = dispersion_parameter(obj.rr.d) ? obj.pp.chol.rank + 1 : obj.pp.chol.rank
+
+dof_residual(obj::GeneralizedLinearModel) = nobs(obj) - dof(obj) + 1
+
```
I think this is redundant with the definition for `LinPredModel`.
```suggestion
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+                   dropcollinear=true)
+
+        @test coef(mdl1) ≈ coef(mdl2)
+        @test stderror(mdl1)[1:3] ≈ stderror(mdl2)[1:3]
```
Also test `NaN` coefficients. Same below.

mousum-github replied:
Added `NaN` test in a test case.

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+        @test deviance(mdl) ≈ 0.2683284457477991
+        @test loglikelihood(mdl) ≈ 0.2177608775670037
+        @test aic(mdl) ≈ 7.564478244866
+        @test predict(mdl) ≈ [14.17008797653959, 13.56744868035191, 24.04398826979472,
+                              19.99413489736071, 11.22434017595308]
```
Can you also check confidence intervals?

mousum-github replied:
okay

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+
+        mdl = glm(@formula(y ~ x1 + x2 + x3), dfrm, Binomial(), LogitLink();
+                   dropcollinear=true)
+        #@test coef(mdl)[1:3] ≈ [-0.0004214049869026, 0.0015203302828698, -0.0015318739665768]
+        #@test stderror(mdl)[1:3] ≈ [0.0063246304255069, 0.0076210032633906, 0.0021094442243717]
```
What's the problem here?

mousum-github replied:
Initially, these two were commented on due to the mismatching of the order of the dropped variable. 
Later, I found the standard errors are not matching more many test cases.

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,85 @@ end
+        @test dof_residual(mdl) == 99998
+        @test aic(mdl) ≈ 138634.84420052
+        @test GLM.dispersion(mdl.model, true) ≈ 1
+    end
```
Also test `predict`?

mousum-github replied:
Okay - we will add a prediction comparison.

mousum-github replied:
Added prediction comparisons in the test case.

nalimilan commented on src/glmfit.jl:
```diff
@@ -273,6 +273,10 @@ end
 
 dof(x::GeneralizedLinearModel) = dispersion_parameter(x.rr.d) ? length(coef(x)) + 1 : length(coef(x))
 
+dof(obj::GeneralizedLinearModel{<:GlmResp,<:DensePredChol{<:Real,<:CholeskyPivoted}}) = dispersion_parameter(obj.rr.d) ? obj.pp.chol.rank + 1 : obj.pp.chol.rank
```
Defining this method for `CholeskyPivoted` here breaks the organization of internal interfaces. Better define `rank(::LinPred)` in linpred.jl for each `LinPred` type and call it here instead of `length(coef(x))` to get the rank for any `GeneralizedLinearModel`.

mousum-github replied:
Okay.

mousum-github replied:
I tried defining a `rank` function, which is conflict with the rank function of linear algebra. So added a `model_rank`.
I am sure there might be a way to overload the `rank` function. Let me know your view.

nalimilan replied:
`model_rank` is OK. But I was thinking about defining `rank(::LinPred)`, not `rank(::GeneralizedLinearModel)`. Otherwise it doesn't allow organizing interfaces more logically (by defining `rank(::LinPred)` in linpred.jl).

nalimilan replied:
To be clearer: I think you can drop `model_rank(::GeneralizedLinearModel)` and `model_rank(::LinearModel)` and just call `model_rank(x.pp)` directly instead.

BTW, thinking about it, maybe `linpred_rank` would be more appropriate than `model_rank`? Or just override the `LinearAlgebra` function using `LinearAlgebrarank(::LinearModel) = ...`.

nalimilan reviewed (commented):
Thanks for taking this up! Sorry for the delay...

ViralBShah commented:
@mousum-github Can you fix the failing doctests?

mousum-github commented:
Hi @nalimilan, 
Covered most of your suggestions. But there is a conflict in `glmfit.jl` file and unable check whether all checks are passed or not.
Sorry for the delay in response. 
cc: @ViralBShah

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
 end
+
+@testset "dropcollinear with GLMs" begin
+
+    data = DataFrame(x1 = [4, 5, 9, 6, 5], x2 = [5, 3, 6, 7, 1], 
+                     x3=[4.2, 4.6, 8.4, 6.2, 4.2], y=[14, 14, 24, 20, 11]
+                    )
```
```suggestion
    data = DataFrame(x1=[4, 5, 9, 6, 5], x2=[5, 3, 6, 7, 1], 
                     x3=[4.2, 4.6, 8.4, 6.2, 4.2], y=[14, 14, 24, 20, 11])
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+                   dropcollinear=true)
+
+        @test coef(mdl1) ≈ coef(mdl2)
+        @test stderror(mdl1)[1:3] ≈ stderror(mdl2)[1:3]
```
```suggestion
        @test stderror(mdl1)[1:3] ≈ stderror(mdl2)[1:3]
        @test isnan(stderror(mdl1)[4])
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+    dfrm.y = Int.(randn(StableRNG(12345), num_rows) .> 0)
+
+    @testset "Test Logistic Regression Outputs from R" begin
+
+        mdl = glm(@formula(y ~ x1 + x2 + x3), dfrm, Binomial(), LogitLink();
+                   dropcollinear=true)
```
```suggestion
        mdl = glm(@formula(y ~ x1 + x2 + x3), dfrm, Binomial(), LogitLink();
                  dropcollinear=true)
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+        mdl = glm(@formula(y ~ x1 + x2 + x3), dfrm, Binomial(), LogitLink();
+                   dropcollinear=true)
+        @test coef(mdl) ≈ [0.005134641298499415, 0.001277642634696454, 0, -0.001347197592412351] atol = 1.0E-8
+        _stdrr = stderror(mdl)
+        @test isnan(_stdrr[3]) == true
+        @test vcat(_stdrr[1:2], _stdrr[4])  ≈ [0.006324636330358, 0.007607056717909, 0.002108547332981] atol = 1.0E-7
```
```suggestion
        stderr = stderror(mdl)
        @test isnan(stderr[3]) == true
        @test vcat(stderr[1:2], stderr[4])  ≈ [0.006324636330358, 0.007607056717909, 0.002108547332981] atol = 1.0E-7
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+        @test aic(mdl) ≈ 138634.32329047
+        @test GLM.dispersion(mdl.model, true) ≈ 1
+        @test predict(mdl)[1:3] ≈ [0.5008036876837788, 0.5001329059586855, 0.5034217691502286]
+        @test confint(mdl)[1:2,1:2] ≈ [-0.007261391766554954 0.017530806817981207;
+                                   -0.013632009007684477 0.016187394900004416] atol = 1.0E-3
```
```suggestion
        @test confint(mdl)[1:2,1:2] ≈ [-0.007261391766554954 0.017530806817981207;
                                       -0.013632009007684477 0.016187394900004416] atol = 1.0E-3
```

nalimilan replied:
Also test other values in the result of `confint`.

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+    @testset "Test with Outputs from R" begin
+        mdl = glm(@formula(y ~ x1 + x2 + x3), data, Normal(), IdentityLink();
+                   dropcollinear=true)
+        @test coef(mdl)[1:3] ≈ [1.350439882697950, 1.740469208211143, 1.171554252199414]
```
```suggestion
        @test coef(mdl)] ≈ [1.350439882697950, 1.740469208211143, 1.171554252199414, 0.0]
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+                     x3=[4.2, 4.6, 8.4, 6.2, 4.2], y=[14, 14, 24, 20, 11]
+                    )
+
+    @testset "Test with equivalent LM model" begin
```
```suggestion
    @testset "Check normal with identity link against equivalent linear model" begin
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+        @test aic(mdl1) ≈ aic(mdl2)
+        @test predict(mdl1) ≈ predict(mdl2)
+    end
+    @testset "Test with Outputs from R" begin
```
```suggestion
    @testset "Check normal with identity link against outputs from R" begin
```

nalimilan commented on test/runtests.jl:
```diff
@@ -1314,3 +1314,66 @@ end
+                              19.99413489736071, 11.22434017595308]
+    end
+
+    num_rows = 100_000
```
Sorry for not spotting this earlier, but can you use a lower number of observations? There's no need to use so much memory here, 1000 observations should be enough.

mousum-github replied:
Okay.

nalimilan commented:
Thanks. I can have a look at the conflict once the PR is ready if you can't fix it. Just ensure tests pass on your side for now.

nalimilan commented:
I've a branch which fixes the conflict with master but I can't push the changes to your branch. Can you give me the necessary permissions? See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork. Otherwise you can do `git fetch && git cherry-pick 998393f9d31b29a3bde77373c2fbbfd8844353cb` to apply my merge commit to your branch.

ayushpatnaikgit commented:
> I've a branch which fixes the conflict with master but I can't push the changes to your branch. Can you give me the necessary permissions? See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork. Otherwise you can do `git fetch && git cherry-pick 998393f9d31b29a3bde77373c2fbbfd8844353cb` to apply my merge commit to your branch.

Hi Milan, 
Since we have done the PR from an organisation account, GitHub doesn't permit us to allow edits from maintainers. Can we please add you as an outside collaborator to xKDR?

ViralBShah commented:
May I suggest we give @mousum-github commit access here, with the usual rules of working through PRs like we have everyone else?

bkamins commented on src/glmfit.jl:
```diff
@@ -521,6 +528,10 @@ const FIT_GLM_DOC = """
+    - `dropcollinear::Bool=true`: Controls whether or not `lm` accepts a model matrix which
+      is less-than-full rank. If `true` (the default), only the first of each set of
+      linearly-dependent columns is used. The coefficient for redundant linearly dependent columns is
+      `0.0` and all associated statistics are set to `NaN`.
```
this statement does not seem to be precise. Consider e.g. the following. `x1, x2, x3, x4, x5` columns (in this order) where `x1`, `x2`, and `x4` are generated randomly and independent, and then `x3=x1+x2` and `x5=x3+x4`.

The implementation drops columns `x3` and `x4`. Reading the description I would expect `x3` and `x5` to be dropped.

@nalimilan - this seems to be unrelated to this PR but a general issue (and probably is a detail that can be resolved later)

nalimilan replied:
Good catch. Do you think it's clearly defined what columns will be dropped by the pivoted cholesky decomposition? In your example, why is `x4` dropped and not `x5`?

bkamins replied:
I think it would be also useful to test the case when collinearity is introduced by `@formula` term, e.g. `@formula(y ~ identity(x1+x2)+x1+x2)` (I think nothing bad should happen but I would test it)

bkamins commented on test/runtests.jl:
```diff
@@ -1446,3 +1446,64 @@ end
     end
 end
+
+@testset "dropcollinear with GLMs" begin
```
where is `dropcollinear=false` tested?

mousum-github replied:
Thanks, @bkamins. I will add test cases accordingly.

bkamins commented:
> May I suggest we give @mousum-github commit access here

This will not resolve the issue. It is the reverse that is needed (i.e. @nalimilan needs to have access to PR source repository).

---

Regarding commit access (as it is an independent issue).

If someone from xKDR team (or the whole team) were willing to work on improving GLM.jl in long term it would be really great. There are many open issues in GLM.jl currently + there are loads of things that could be added.

However, is that it is not the person who makes a PR, but a person who reviews it that needs commit access.
Currently mostly @nalimilan serves as a reviewer on this repository (and @nalimilan - thank you for that).
Of course we could change that in the future (there are many other packages that need attention by @nalimilan so having someone else oversee GLM.jl after some transition period, during which the person would learn JuliaStats standards, would be great).

Given these considerations @mousum-github - would you find having commit access to GLM.jl useful?

ViralBShah commented:
> > May I suggest we give @mousum-github commit access here
> 
> This will not resolve the issue. It is the reverse that is needed (i.e. @nalimilan needs to have access to PR source repository).
> 
> Regarding commit access (as it is an independent issue).
> 
> If someone from xKDR team (or the whole team) were willing to work on improving GLM.jl in long term it would be really great. There are many open issues in GLM.jl currently + there are loads of things that could be added.
> 
> However, is that it is not the person who makes a PR, but a person who reviews it that needs commit access. Currently mostly @nalimilan serves as a reviewer on this repository (and @nalimilan - thank you for that). Of course we could change that in the future (there are many other packages that need attention by @nalimilan so having someone else oversee GLM.jl after some transition period, during which the person would learn JuliaStats standards, would be great).
> 
> Given these considerations @mousum-github - would you find having commit access to GLM.jl useful?

Right - my suggestion was so that @mousum-github can then make PRs (perhaps even this one) using branches on this repo. That would make it easier for @nalimilan and everyone to edit directly and what not. And this would also serve as a first step towards having more contributors such as @mousum-github overseeing this repo.

mousum-github commented:
I am sure that the commit access will be very useful, and I am happy to help maintain the project going forward.

nalimilan commented:
Thanks @ayushpatnaikgit. Not sure why GitHub doesn't support that, but thanks to the access to the xKDR repo I could push my merge commit. Tests still appear to fail though.

Usually we only give commit access to people who have made several PRs over a relatively long period, but given that xKDR is invested in improving the stats ecosystem I guess we can accelerate that process. I've just sent an invitation to @mousum-github. Help reviewing other PRs would of course be welcome!

mousum-github commented:
> Thanks @ayushpatnaikgit. Not sure why GitHub doesn't support that, but thanks to the access to the xKDR repo I could push my merge commit. Tests still appear to fail though.
> 
> Usually we only give commit access to people who have made several PRs over a relatively long period, but given that xKDR is invested in improving the stats ecosystem I guess we can accelerate that process. I've just sent an invitation to @mousum-github. Help reviewing other PRs would of course be welcome!

Thanks @nalimilan

mousum-github commented:
Hi @nalimilan,
We have updated the code and documentation.

We have updated the newly added `nulldeviance` and `nullloglikelihood` functions in `glm`. Passed the `dropcollinear` argument value to `fit` function for null models. 

We also added a test case to check the `PosDefException` exception when `dropcollinear` is false for a rank-deficient model. Unfortunately, the test case was not passed in a few systems.  So, we have commented the test case for timing.

nalimilan commented:
> We also added a test case to check the `PosDefException` exception when `dropcollinear` is false for a rank-deficient model. Unfortunately, the test case was not passed in a few systems. So, we have commented the test case for timing.

Maybe another example would pass tests more reliably? We really need to check that path. How about trying with categorical variables, like in the "rankdeficient" testset for linear models?

mousum-github commented:
> 

Okay. Let me check

mousum-github commented:
> > 
> 
> Okay. Let me check.

Added two test cases similar to `rankdeficient`.

bkamins commented on test/runtests.jl:
```diff
@@ -19,6 +19,155 @@ function simplemm(x::AbstractVecOrMat)
+        @test predict(mdl1) ≈ predict(mdl2)
+    end
+
+    #@testset "Check `PosDefException` handling when dropcollinear = false" begin
```
why is this commented out?

mousum-github replied:
We added the test case to check the PosDefException exception when dropcollinear is false for a rank-deficient model. Unfortunately, the test case was not passed in a few systems. So, we have commented the test case for timing.
Later, we added two similar test cases.

nalimilan reviewed (commented):
Seems almost ready! Just a few stylistic comments.

mousum-github commented on test/runtests.jl:
```diff
@@ -19,6 +19,155 @@ function simplemm(x::AbstractVecOrMat)
+        @test isapprox(deviance(m1), 0.12160301538297297)
+        Xmissingcell = X[inds, :]
+        ymissingcell = y[inds]
+        @test_throws PosDefException m2 = fit(GeneralizedLinearModel, Xmissingcell, ymissingcell, Normal(); dropcollinear=false)
```
Okay.

mousum-github commented on test/runtests.jl:
```diff
@@ -19,6 +19,155 @@ function simplemm(x::AbstractVecOrMat)
+    #    @test_throws PosDefException mdl2 = glm(@formula(y ~ x1 + x2 + x3), data, Normal(), IdentityLink();
+    #               dropcollinear=false)
+    #end
+
```
Removed the test case

bkamins commented on src/glmfit.jl:
```diff
@@ -521,6 +526,10 @@ const FIT_GLM_DOC = """
 
     # Keyword Arguments
+    - `dropcollinear::Bool=true`: Controls whether or not `lm` accepts a model matrix which
+      is less-than-full rank. If `true` (the default), only the first of each set of
+      linearly-dependent columns is used. The coefficient for redundant linearly dependent columns is
+      `0.0` and all associated statistics are set to `NaN`.
```
```suggestion
      is less-than-full rank.
      If `true` (the default) the coefficient for redundant linearly dependent columns is
      `0.0` and all associated statistics are set to `NaN`.
      Typically from a set of linearly-dependent columns is the last ones are identified as redundant
      (however, the exact selection of columns identified as redundant is not guaranteed).
```
@nalimilan - per our discussion, maybe we should change the docstring as I propose here?

nalimilan reviewed (approved):
Thanks!
